crates/subspace-farmer/src/farmer_cache.rs

```diff
@@ -217,139 +231,90 @@ where
         &self,
         piece_getter: &PG,
         worker_state: &mut CacheWorkerState,
-        new_piece_caches: Vec<Arc<dyn PieceCache>>,
+        mut new_piece_caches: Vec<Arc<dyn PieceCache>>,
     ) where
         PG: PieceGetter,
     {
         info!("Initializing piece cache");
-        // Pull old cache state since it will be replaced with a new one and reuse its allocations
-        let cache_state = mem::take(&mut *self.piece_caches.write().await);
-        let mut stored_pieces = Vec::with_capacity(new_piece_caches.len());
-        let mut free_offsets = Vec::with_capacity(new_piece_caches.len());
-        for mut state in cache_state {
-            state.stored_pieces.clear();
-            stored_pieces.push(state.stored_pieces);
-            state.free_offsets.clear();
-            free_offsets.push(state.free_offsets);
+        const MAX_CACHES_NUM: usize = u16::MAX as usize;
+        if new_piece_caches.len() > MAX_CACHES_NUM {
+            warn!(
+                piece_caches_number = ?new_piece_caches.len(),
+                "Too many piece caches provided; \
+                Any extra caches beyond the {MAX_CACHES_NUM} limit will be ignored",
+            );
+            new_piece_caches.truncate(MAX_CACHES_NUM);
         }
-        stored_pieces.resize(new_piece_caches.len(), HashMap::default());
-        free_offsets.resize(new_piece_caches.len(), VecDeque::default());
+
+        let mut backends = Vec::new();
+        #[allow(clippy::mutable_key_type)]
+        let mut stored_pieces = HashMap::new();
+        let mut free_offsets = VecDeque::new();
```

Comment:
Please do not do `::new()` if you can avoid it, it will result in bad performance and bad memory usage. See how previous code was trying to carefully preallocate all data structures to correct size beforehand, in fact it was even reusing previous memory allocations rather than doing fresh allocations. I think it might be good to preserve that.